### PR TITLE
Summer cleanup

### PR DESCRIPTION
**Pull request by drdanz**
Mostly cleanup licenses and docs, move implementations, in `.cpp` files (`-inl.h` files for templates).

The only important change is that c++11 is now required also in the .h files, therefore the compiler must support c++11 also to use YARP (the required flags are automatically enabled)

### CODE REVIEW

File: src/libYARP_OS/CMakeLists.txt [hunk 1]
```diff
@@ -393,6 +404,10 @@ target_include_directories(YARP_OS PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOUR
 target_link_libraries(YARP_OS PRIVATE ${ACE_LIBRARIES}
                               PUBLIC YARP::YARP_conf)
 
+target_compile_features(YARP_OS PUBLIC cxx_nullptr
+                                       cxx_override
+                                       cxx_rvalue_references)
```
Comment (traversaro) [inline src/libYARP_OS/CMakeLists.txt]: `target_compile_features` is a CMake 3.1 command, so we should increase the `cmake_minimum_required` version to CMake 3.1 . Furthermore, as we still need to support Debian Jessie for the pc104 of several iCub robots, we should update the pc104 image (http://wiki.icub.org/wiki/The_Linux_on_the_pc104) to include CMake 3.6 from `jessie-backports`.

Comment (drdanz) [inline src/libYARP_OS/CMakeLists.txt]: 💩 💩 💩 💩 💩 💩 💩 💩 💩 
I forgot!

File: src/libYARP_OS/CMakeLists.txt [hunk 2]
```diff
@@ -404,9 +404,18 @@ target_include_directories(YARP_OS PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOUR
...
+endif()
+if(CMAKE_VERSION VERSION_LESS 3.1)
+  if(DEFINED CXX11_FLAGS)
+    target_compile_options(YARP_OS PUBLIC ${CXX11_FLAGS})
+  endif()
+else()
+  target_compile_features(YARP_OS PUBLIC cxx_nullptr
+                                         cxx_override
+                                         cxx_rvalue_references)
+endif()
```
Comment (traversaro) [inline src/libYARP_OS/CMakeLists.txt]: @diegoferigo @francesco-romano If we want to start using C++11 feature in iDynTree headers we should add this code to `idyntree-core` .

Comment (drdanz) [inline src/libYARP_OS/CMakeLists.txt]: btw, not sure if in cmake 3.8 or 3.9 there is the `cxx_std_11` feature that simplifies this... I don't think it's worth to have an extra `if` here though...

**Pull request merged**